docs/user-guide/reduction-workflow-guidelines.md

```diff
@@ -0,0 +1,139 @@
+# Reduction Workflow Guidelines
+
+## Introduction
+
+TODO
+
+- Package and module structure (where to place types? what goes where?)
+  - Requires minor discussion
+- Loading from SciCat vs. local files (e.g., define run ID, choose provider that either converts to local path, or uses service to get file and return path)
+- Should we have default params set in workflows?
+  - Avoid unless good reason.
+  - Can have widgets that generate dict of params and values, widgets can have defaults
+
+## Nomenclature
+
+- *Provider*: A callable step in a workflow writing with Sciline.
+
+## D: Documentation
+
+### D.1: Document math and references in docstrings
+
+**Reason**
+Documentation should be as close to the code as possible, to decrease the chance that it runs out of sync.
+This includes mathematical formulas and references to literature.
+
+**Note**
+We have previously documented math and references in Jupyter notebooks.
+This is not sufficient, as the documentation is not close to the code.
+
+## N: Naming
+
+TODO
+- Naming conventions (and type conventions (example: filenames)?)
+
+## P: Performance
+
+### P.1: Runtime and memory use of workflows shall be tested with large data
+
+**Reason**
+We want to ensure that the workflows are efficient and do not consume excessive memory.
+
+**Note**
+This is often not apparent from small test data, as the location of performance bottlenecks may depend on the size of the data.
+
+## S: Structure
+
+### S.1: Workflows shall be able to return final results as event data
+
+**Reason**
+- Required for polarization analysis, which wraps a base workflow.
+- Required for subsequent filtering, unless part of the workflow.
+
+**Note**
+There should be a workflow parameters (flag) to select whether to return event data or not.
+
+### S.2: Split loading and handling of monitors and detectors from loading of auxiliary data and metadata
+
+**Reason**
+- Allows for more efficient parallelism and reduction in memory use.
+- Avoids loading large data into memory that is not needed for the reduction.
+- Avoids keeping large data alive in memory if output metadata extraction depends on auxiliary input data or input metadata.
+
+### S.3: Avoid dependencies of output metadata on large data
+
+**Reason**
+Adding dependencies on large data to the output metadata extraction may lead to large data being kept alive in memory.
+
+**Note**
+Most of this is avoid by following S.2.
+A bad exampleAvoid would be to write to total raw counts to the output metadata, as this would require keeping the large data alive in memory, unless it is ensured that the task runs early.
```

Comment:
But how do we do this instead? This note just forbids writing certain things as metadata. But that cannot be avoided, e.g., in reflectometry as we need to write wavelength and theta ranges.

Reply:
There will be exceptions, not everything can be "avoided", this is why I chose this formulation instead of making it stricter.

Reply:
Then I'd like clarification about how strict these guidelines are. As it is written, it sounds like a command. So if this one may be violated, can all rules be broken if we can argue for it?

Reply:
Time will tell? I think we should be agile and avoid over-analyzing for now, see what works in practice and what does not, then iterate and improve.

Reply:
Also, in the case you mentioned, shouldn't the wavelength and theta ranges report the *used* ranges, instead of those in the raw data? I presume the workflow may limit this, and in any case it can be computed from derived data instead of raw data, limiting the problem.

Reply:
> Also, in the case you mentioned, shouldn't the wavelength and theta ranges report the _used_ ranges, instead of those in the raw data? I presume the workflow may limit this, and in any case it can be computed from derived data instead of raw data, limiting the problem.

Possibly. The current workflow does not allow limiting theta. Also, what if the bounds are much larger than the data? I'd want to report the limits of the data that is actually used instead.